### ISSUE TITLE
[no-this-alias] should only be flagged for `const` initialization

### ISSUE THREAD

**Issue by dinofx** [typescript-eslint/typescript-eslint#2576]
A variable initially assigned to `this` but which gets reassigned later to something else shouldn't be flagged.

Here's an example method in a class:

```TS
  private getPath() {
    const result: string[] = [];
    let n: JSONNode | undefined = this;
    do {
      result.unshift(n.label);
      n = n.parent;
    } while (n);
    return result;
  }
```

This rule should only flag a constant assigned to `this`.

**Comment by bradzacher** [comment]
Could you explain why you think it's any different to allow `let` with reassignment?

It seems to me like regardless of if it remains that way, `n` is at some point an alias for `this`.

**Comment by dinofx** [comment]
Is the title (should only be flagged for `const` initialization), I was assuming that `prefer-const` rule was being used, in which case a `let` is only used if the variable is reassigned.  But to state it more generally, the rule should only flag variables that are assigned to `this` and never reassigned.

**Comment by bradzacher** [comment]
I disagree.
You have aliased `this` into the variable `n`.
If you later reassign it, that doesn't change the fact that `n` was an alias for `this`.

**Comment by dinofx** [comment]
From the docs:
_"Assigning a variable to this instead of properly using arrow lambdas"_

Really, this rule should be called `use-arrow-functions-to-access-this-from-the-enclosing-scope`

If a variable is going to later point to something that is not `this`, then there is no way to use the recommended practice (arrow lambdas) as an alternative.

The fact that the variable is reassigned changes everything.  The variable is **not** an alias for `this`.  It temporarily pointed to this, but later to some other object.

**Comment by bradzacher** [comment]
Given that nobody else has engaged with this - I don't think there's enough community need for it.
So I'm going to close this as not going to implement.
